### ISSUE TITLE
neural network liquids

### ISSUE THREAD

**Issue by PavelBlend** [ProjectPhysX/FluidX3D#178]
@ProjectPhysX Hello.

Is it possible to train a neural network so that it can calculate liquids? So that the program "guesses" all the attributes of the voxels (pressure, speed, etc.). Or does this task require too much time and resources? If you rewrite your program using neural networks, will it run faster?

**Comment by ProjectPhysX** [comment]
Hi @PavelBlend,

a lot of hype has recently gone into AI fluid simulations, but for most applications this is a dead end.
A neural network only does interpolation between the known good solutions it was trained on and tries extrapolation outside of the known parameter space which often fails. It is certainly possible to train a NN on solutions of fluid simulations, and it will learn the typical flow behaviour for specific application cases. There are big caveats however:
- Fluid simulations are a super large parameter space (viscosity from Stokes flow to turbulence, boundary geometries, free surface behavior etc. Covering all this in training is impossible, because each training set sample is a full-blown real CFD simulation. Only a very small part of the parameter space can be covered at a time, which makes AI fluid simulations very specific and not general purpose.
- Training is excessively expensive, about 4 orders of magnitude more energy than just running one CFD simulation directly, and needs to be done for every new application.
- AI simulations themselves are slower than LBM simulations. CFD is a problem of data movement. The less data has to be copied to update a grid cell's state in VRAM, the faster. An explicit physical solver like LBM loads the previous cell state, does proper math, and stores the result in VRAM - for each cell independently and in parallel, and with a minimal amount of memory transactions. FluidX3D does this in 77 Bytes/cell per time step. On top, CFD solvers have the math nailed down - NNs try to emulate it with only matrix-vector multiplications and activation functions, which needs more arithmetic operations.
- Transient simulations are entirely out of scope for AI methods. The grid cell data has to be stored somewhere in memory, and AI can't reduce the memory footprint either.

I've only seen one good use-case of AI CFD so far: At DLR they optimized turbine blade geometry. This is in transsonic regime, where LBM-type methods fail and much slower implicit CFD solvers are needed. They trained a NN on results of ~200 CFD supercomputer simulations with slightly different blade geometries, and used the NN to interpolate in the data to find an optimal blade geometry. This is more of a linking between turbine blade geometry parameter space and results parameter space, without actually AI-simulating the physics in between. Find a good video about this [here](https://youtu.be/puygJ11vfdk) (in German though).

Kind regards,
Moritz
